### PR TITLE
WIP Failing test for try-with-resources

### LINKED ISSUE #46: Add test for try-with-resources statement
Enhance the test suite to verify that variables in tests containing `try-with-resource` statements are correctly recorded. Similar tests can be found in the [junit4-tests](https://github.com/dodie/scott/tree/master/scott-tests/junit4-tests) project.

Also, add a failing test to the [showcase project](https://github.com/dodie/scott/tree/master/scott-examples/junit4), to make sure that the rendered failure message looks pretty.

Note: please make sure to visit the [contribution](https://github.com/dodie/scott/tree/master/CONTRIBUTING.md) and [development guide](https://github.com/dodie/scott/tree/master/docs/development-guide.md) for some important notes on how to build and debug Scott.

### PR DESCRIPTION
**Pull request by kevinkraus**
Closes #46

### CODE REVIEW

**Comment by kevinkraus** [comment]
I'm still working on the `EmptyStackException` that occurs in this test; please don't merge this yet.

**Comment by dodie** [comment]
Hey @kevinkraus !

How's it going? Can I help you with anything? In the meantime, make sure to read the [developer guide](https://github.com/dodie/scott/blob/master/docs/development-guide.md) if you haven't already. I know it's not much, but maybe it can be a good starting point.

Cheers,
David

**Comment by kevinkraus** [comment]
Hey @dodie - yes, I think I could use some help getting pointed in the right direction. First off, I'm relatively new to bytecode instrumentation :) it looks like `ScopeExtractorTestMethodVisitor .visitTryCatchBlock()` gets called 10 times for my test, where it only gets called once for a standard try/catch block. 

Because of that, the collection `tryCatchBlocks` [here](https://github.com/dodie/scott/blob/master/scott/src/main/java/hu/advancedweb/scott/instrumentation/transformation/ScopeExtractorTestMethodVisitor.java#L149) has 10 entries instead of the 1 that it should have. 

I'm not really following why (it appears) that `visitTryCatchBlock()` gets called 10 times. Any ideas that could point me in the right direction?

**Comment by kevinkraus** [comment]
By the way @dodie, this is a very nice codebase!

**Comment by dodie** [comment]
Hey @kevinkraus, thanks! :beers: 

When running your test I experience a quite different error, but I think it's related to the same bug. 

I added this test to `hu/advancedweb/example/CounterTest` in the `scott-examples/junit4/` module: 

```java
@Test
public void try_with_resources() throws IOException {
    try (ByteArrayOutputStream out = new ByteArrayOutputStream();
         PrintStream printStream = new PrintStream(out)) {
         printStream.append("Hello World");
    }
}
```

When I try to execute it with `mvn install`, the JVM crashes with the following message:
```
[ERROR] java.lang.VerifyError:
[ERROR] Bad local variable type
[ERROR] Exception Details:
[ERROR] Location:
[ERROR] hu/advancedweb/example/CounterTest.try_with_resources()V @246: aload
[ERROR] Reason:
[ERROR] Type top (current frame, locals[4]) is not assignable to reference type
[ERROR] Current Frame:
[ERROR] bci: @246
[ERROR] flags: { }
[ERROR] locals: { 'hu/advancedweb/example/CounterTest', 'java/lang/Throwable', 'java/lang/Throwable', 'java/io/ByteArrayOutputStream' }
[ERROR] stack: { 'java/io/ByteArrayOutputStream' }
[ERROR] Bytecode:
[ERROR] 0000000: 1255 1270 b800 5c01 4c01 4dbb 002b 59b7
[ERROR] 0000010: 002d 4e12 7112 7212 7312 70b8 0063 2d12
[ERROR] 0000020: 7212 7312 70b8 0067 bb00 2e59 2d2d 1272
[ERROR] 0000030: 1273 1270 b800 67b7 0030 3a04 1274 1275
[ERROR] 0000040: 1276 1270 b800 6319 0412 7512 7612 70b8
[ERROR] 0000050: 0067 1904 1233 2d12 7512 7312 70b8 0067
[ERROR] 0000060: 1904 1275 1276 1270 b800 67b6 0035 5719
[ERROR] 0000070: 04c6 0042 1904 2d12 7712 7312 70b8 0067
[ERROR] 0000080: 1904 1277 1276 1270 b800 67b6 0039 a700
[ERROR] 0000090: 254c 1904 c600 1d19 042d 1278 1273 1270
[ERROR] 00000a0: b800 6719 0412 7812 7612 70b8 0067 b600
[ERROR] 00000b0: 392b bf2d c600 682d 2d12 7812 7312 70b8
[ERROR] 00000c0: 0067 1904 1278 1276 1270 b800 67b6 003c
[ERROR] 00000d0: a700 4c4d 2bc7 0008 2c4c a700 0d2b 2ca5
[ERROR] 00000e0: 0008 2b2c b600 3d2d c600 1c2d 2d12 7812
[ERROR] 00000f0: 7312 70b8 0067 1904 1278 1276 1270 b800
[ERROR] 0000100: 67b6 003c 2bbf 4d2b c700 082c 4ca7 000d
[ERROR] 0000110: 2b2c a500 082b 2cb6 003d 2bbf b1
[ERROR] Exception Handler Table:
[ERROR] bci [82, 111] => handler: 145
[ERROR] bci [40, 179] => handler: 211
[ERROR] bci [11, 262] => handler: 262
[ERROR] Stackmap Table:
[ERROR] full_frame(@145,{Object[#1],Null,Null,Object[#43],Object[#46]},{Object[#62]})
[ERROR] full_frame(@177,{Object[#1],Object[#62],Null,Object[#43],Object[#46]},{})
[ERROR] full_frame(@179,{Object[#1],Null,Null,Object[#43],Object[#46]},{})
[ERROR] full_frame(@211,{Object[#1],Object[#62],Null,Object[#43]},{Object[#62]})
[ERROR] full_frame(@221,{Object[#1],Object[#62],Object[#62],Object[#43]},{})
[ERROR] same_frame(@231)
[ERROR] same_frame(@260)
[ERROR] full_frame(@262,{Object[#1],Object[#62],Object[#62]},{Object[#62]})
[ERROR] same_frame(@272)
[ERROR] same_frame(@282)
[ERROR] full_frame(@284,{Object[#1],Null,Null,Object[#43],Object[#46]},{})
```

It's just a guess, but I think the code you've linked fails to properly determine the variable scopes when it comes to `try-with-resources`. Scott tracks variable states for each operation (e.g. method calls), for variables that are in scope. I think the generated `close` call for the `try-with-resources` triggers Scott to record variables, and it tries to access variables that are out of scope.

I think you've correctly found the source of the problem. It seems that the try-catch handling is not implemented correctly. When I wrote that code I did not really investigate how try-catch statements manifests themselves in bytecode, and my assumption was that there will be one instruction that describes the whole statement. It seems I was wrong. The generated bytecode contains multple entries in the ```Exception table```, that contains the range that it guards, the location of the handler code and the exception type. It also generates an entry without an exception type for the `finally` block. In addition to that, `try-with-resources` generates  a"finally" block for each resource it manages.

The `visitTryCatchBlock` method gets called for each entry in the exception table, and this seems to cause some problems.

Based on this, I think the problem applies to regular finally blocks, or blocks with multiple catch statements, not just the ones generated by `try-with-resources`.

To find out what's going on, I'd suggest the following:

- Check out the generated bytecode of various test methods using `javap`. Compare the bytecode produced by a simple try-catch (multiple catches with/without finally) with the bytecode of a `try-with-resources`.
For example:
```shell
mvn install -DskipTests
cd target/test-classes
javap -c hu.advancedweb.example.CounterTest
```

- I think it's useful to have the ASM Bytecode Outliner plugin installed. It's pretty awesome: as you navigate in the code, it displays the bytecode of the selected method and class. I use Eclipse, so I can speak [for that version](https://marketplace.eclipse.org/content/bytecode-outline), but there is a plugin for IntelliJ as well. (Note: the entries for the exception table appear as `TRYCATCHBLOCK` instructions.)

An image speaks a thousand words:
image

Sorry for the lengthy answer, I hope this might help.

Cheers,
David

**Comment by kevinkraus** [comment]
This is great information, thanks for the guidance! I'll play with this for a few days and see what I can do.

**Comment by kevinkraus** [comment]
Quick update - I'm still working through understanding the differences between the try-with-resources and the traditional try-catch blocks. I have no questions as of now, just exploring.

**Comment by kevinkraus** [comment]
Ok, I give up on IntelliJ's bytecode plugins ;) I'm installing Eclipse

**Comment by kevinkraus** [comment]
@dodie do you follow why the handlers (L2, L5, L7) don't have line numbers associated with them? I do see references to the `close()` methods that will be called as part of the `try-with-resources` in those labels. Also, this bytecode from the Eclipse plugin and the `javap` output both have 3 entries in the exception table, which is less than I'd expect based on your notes (and different than what IntelliJ outputs) - any thoughts on that?

image

```
 public try_with_resources()V throws java/io/IOException 
  @Lorg/junit/Test;()
    TRYCATCHBLOCK L0 L1 L2 
    TRYCATCHBLOCK L3 L4 L5 
    TRYCATCHBLOCK L6 L7 L7 
   L8
    LINENUMBER 16 L8
    ACONST_NULL
    ASTORE 1
    ACONST_NULL
    ASTORE 2
   L6
    NEW java/io/ByteArrayOutputStream
    DUP
    INVOKESPECIAL java/io/ByteArrayOutputStream.<init>()V
    ASTORE 3
   L3
    LINENUMBER 17 L3
    NEW java/io/PrintStream
    DUP
    ALOAD 3
    INVOKESPECIAL java/io/PrintStream.<init>(Ljava/io/OutputStream;)V
    ASTORE 4
   L0
    LINENUMBER 19 L0
    ALOAD 4
    LDC "Hello World"
    INVOKEVIRTUAL java/io/PrintStream.append(Ljava/lang/CharSequence;)Ljava/io/PrintStream;
    POP
   L9
    LINENUMBER 21 L9
    LDC "out"
    INVOKESTATIC hu/advancedweb/scott/TestHelper.getLastRecordedStateForVariable(Ljava/lang/String;)Ljava/lang/String;
    ALOAD 3
    INVOKEVIRTUAL java/io/ByteArrayOutputStream.toString()Ljava/lang/String;
    INVOKESTATIC org/hamcrest/CoreMatchers.equalTo(Ljava/lang/Object;)Lorg/hamcrest/Matcher;
    INVOKESTATIC org/junit/Assert.assertThat(Ljava/lang/Object;Lorg/hamcrest/Matcher;)V
   L1
    LINENUMBER 22 L1
    ALOAD 4
    IFNULL L4
    ALOAD 4
    INVOKEVIRTUAL java/io/PrintStream.close()V
    GOTO L4
   L2
   FRAME FULL [hu/advancedweb/scott/TryWithResourcesRecordingTest java/lang/Throwable java/lang/Throwable java/io/ByteArrayOutputStream java/io/PrintStream] [java/lang/Throwable]
    ASTORE 1
    ALOAD 4
    IFNULL L10
    ALOAD 4
    INVOKEVIRTUAL java/io/PrintStream.close()V
   L10
   FRAME CHOP 1
    ALOAD 1
    ATHROW
   L4
   FRAME SAME
    ALOAD 3
    IFNULL L11
    ALOAD 3
    INVOKEVIRTUAL java/io/ByteArrayOutputStream.close()V
    GOTO L11
   L5
   FRAME SAME1 java/lang/Throwable
    ASTORE 2
    ALOAD 1
    IFNONNULL L12
    ALOAD 2
    ASTORE 1
    GOTO L13
   L12
   FRAME SAME
    ALOAD 1
    ALOAD 2
    IF_ACMPEQ L13
    ALOAD 1
    ALOAD 2
    INVOKEVIRTUAL java/lang/Throwable.addSuppressed(Ljava/lang/Throwable;)V
   L13
   FRAME SAME
    ALOAD 3
    IFNULL L14
    ALOAD 3
    INVOKEVIRTUAL java/io/ByteArrayOutputStream.close()V
   L14
   FRAME CHOP 1
    ALOAD 1
    ATHROW
   L7
   FRAME SAME1 java/lang/Throwable
    ASTORE 2
    ALOAD 1
    IFNONNULL L15
    ALOAD 2
    ASTORE 1
    GOTO L16
   L15
   FRAME SAME
    ALOAD 1
    ALOAD 2
    IF_ACMPEQ L16
    ALOAD 1
    ALOAD 2
    INVOKEVIRTUAL java/lang/Throwable.addSuppressed(Ljava/lang/Throwable;)V
   L16
   FRAME SAME
    ALOAD 1
    ATHROW
   L11
    LINENUMBER 23 L11
   FRAME CHOP 2
    RETURN
   L17
    LOCALVARIABLE this Lhu/advancedweb/scott/TryWithResourcesRecordingTest; L8 L17 0
    LOCALVARIABLE out Ljava/io/ByteArrayOutputStream; L3 L14 3
    LOCALVARIABLE printStream Ljava/io/PrintStream; L0 L10 4
    MAXSTACK = 3
    MAXLOCALS = 5
}
```

**Comment by kevinkraus** [comment]
Quick update - I think I'm onto a fix! I'll start pushing code as soon as I'm sure.

**Comment by dodie** [comment]
Hi @kevinkraus ,
I am glad to hear that! Thank you for all your effort you put into this investigation, I really appreciate it.

> why the handlers (L2, L5, L7) don't have line numbers associated with them?

I've found nothing relevant in the JVM or Java spec, but I think it's due to that these handlers are not "real" in a sense like `catch` or `finally` blocks are, as they are generated by the compiler. Since they don't have any corresponding source code, I'd guess this is the reason why the byte-code lacks the line numbers.

> Also, this bytecode from the Eclipse plugin and the javap output both have 3 entries in the exception table, which is less than I'd expect based on your notes (and different than what IntelliJ outputs) - any thoughts on that?

I did not really have time to look to the bottom of this behavior, however I think this section from the JLS might help: [Chapter 14. Blocks and Statements / 14.20.3. try-with-resources](https://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.20.3). In this section there is an equivalent pre Java7 source code for an example that uses try-with-resources that might help sort out the real behavior of this feature.

Note: I've compiled the following method on CentOS/JDK v1.8.0_151-b12:
```
@Test
public void try_with_resources() throws IOException {
    try (ByteArrayOutputStream out = new ByteArrayOutputStream();
         PrintStream printStream = new PrintStream(out)) {
         printStream.append("Hello World");
    }
}
```

And got the following exception table:
```
Exception table:
       from    to  target type
          38    42    45   Class java/lang/Throwable
          22    29    64   Class java/lang/Throwable
          22    29    73   any
          84    88    91   Class java/lang/Throwable
          64    75    73   any
         118   122   125   Class java/lang/Throwable
          10   110   141   Class java/lang/Throwable
          10   110   146   any
         156   160   163   Class java/lang/Throwable
         141   148   146   any
```

When I tried to run it, it crashed with the `EmptyStackException` you encountered.
Did you also test it on CentOS (or RHEL)?

**Comment by kevinkraus** [comment]
You're welcome! Thanks, that does help, and verifies something interesting I'm seeing. 

When I added a log to the `visitLineNumber()` method in `ScopeExtractorTestMethodVisitor`, I noticed that it was visiting the same line multiple times with different `Label`s. 

This code:
image

Creates the following logs:
```
visiting line 16 start: L824318946
visiting line 17 start: L401625763
visiting line 16 start: L930990596
visiting line 19 start: L939047783
visiting line 21 start: L1921595561
visiting line 22 start: L1237514926
visiting line 16 start: L1134517053
visiting line 22 start: L548246552
visiting line 16 start: L2108649164
visiting line 22 start: L83954662
visiting line 23 start: L565760380
```
Note: I'm seeing similar behavior in a `try/catch/finally` block.

So far, I'm experimenting with 
1. iterating over the `lines` as is currently done in `ScopeExtractorTestMethodVisitor.calculateScope()`, but keeping the association of a `line` to more than one `Label`, kind of like a `MultiMap`. 
1. identifying which entries in the `exception table` are meaningful and should be considered inside `calculateScope()`. 

Please let me know if you have any thoughts on this!

**Comment by kevinkraus** [comment]
Also - my compilation is done on MacOS High Sierra.

**Comment by dodie** [comment]
I think both ideas are good!

I was wondering that it might ease the verification of the fix if we add an additional test for an expanded, old version of a test that uses try-with-resources, something like [this](https://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.20.3.1). (Although they should generate similar bytecode, there might be differences, for example regarding the LINENUMBER annotations.)

**Comment by dodie** [comment]
Hi @kevinkraus!

Just wanted to let you know, that based on the previous findings I did an experiment - however, it did not solve the problem, I thought it's worthwhile to share.

I tried to change the `try-catch` block handling in `ScopeExtractorTestMethodVisitor.java`. Specifically, I've changed the [tryCatchBlockScopes](https://github.com/dodie/scott/blob/master/scott/src/main/java/hu/advancedweb/scott/instrumentation/transformation/ScopeExtractorTestMethodVisitor.java#L149) managing code to two separate for loops:

```
for (TryCatchBlockLabels tryCatchBlock : tryCatchBlocks) {
	if (label == tryCatchBlock.start) {
		tryCatchBlockScopes.push(tryCatchBlock);
	}
}

for (TryCatchBlockLabels tryCatchBlock : tryCatchBlocks) {
	if (label == tryCatchBlock.handler) {
		tryCatchBlockScopes.pop();
	}
}
```
The original approach seems a bit suspicious when it comes to handlers without a `LINENUMBER`, or even when multiple handlers are squeezed into the same line. Because the `tryCatchBlockScopes` is a `Set`, the ordering of the for-loop is not guaranteed. Also, because of the `if-else`, only one part of the code executes. My assumption was that this might result in an unfortunate case, where the code tries to call the `pop` method before it would `push` anything to the stack, so I divided the original code to two loops to make sure the `push` always happens before.
I tested it on my Ubuntu setup, and unfortunately it did not solve the JVM crash. Tomorrow I'll try this on a CentOS machine to see if it affects the case where the `EmptyStackException` occurs.

Cheers,
David

**Comment by dodie** [comment]
The change does solve the EmptyStackException on my other env. Now it produces the same JVM crash. I think we are dealing with 2 issues:
- the random access nature of the set causes problems when the try and catch blocks start in the same line (responsible for the EmptyStackException) 
- the try-catch blocks are managed based on LINENUMBERS, which are not present for generated blocks, thus Scott does not notices the start of the handlers. Because of this, Scott might try to log variables which are out of scope (resulting in bytecode verification error)

Would you mind if I'd commit the fix for the first issue to this PR?

**Comment by kevinkraus** [comment]
That would be great, thanks! I've had some trouble making time to continue on the second issue, but this week looks better.

**Comment by dodie** [comment]
Now it the test instrumentation should work, (no more EmptyStackException), and should crash with the bytecode verification when it tries to load the instrumented code. Could you please check that it behaves the same way on your environment?

**Comment by kevinkraus** [comment]
Yes, it crashes while running the build from Terminal and while running tests in IntelliJ. That's great!

**Comment by kevinkraus** [comment]
I just pushed a test that you suggested a while back, but I misunderstood the original request, I think. This test reproduces the JVM crash without using `try-with-resources`, just nested try/catch/finally statements to make sure all classes get closed correctly.

File: scott-tests/junit4-tests/src/test/java/hu/advancedweb/scott/ExceptionTest.java [hunk 1]
```diff
@@ -139,24 +139,31 @@ public void tryFinallyBlockThatThrowsException() {
...
+			} catch (Exception e) {
+				throwable = e;
+				throw e;
+			} finally {
+				assertThat(TestHelper.getLastRecordedStateForVariable("out"), equalTo(out.toString()));
+				if (throwable != null) {
+					try {
+						printStream.close();
+					} catch (Exception e) {
+						throwable.addSuppressed(e);
```
Comment (kevinkraus) [inline scott-tests/junit4-tests/src/test/java/hu/advancedweb/scott/ExceptionTest.java]: Removing this line makes this test pass (although the functionality is changed) @dodie

**Comment by kevinkraus** [comment]
This is definitely making it easier to see what's going on. There's a nested Exception (line 150) whose scope Scott appears to be a little mixed up about (start line 158, end line 152).

**Comment by dodie** [comment]
Hi @kevinkraus !
Thanks for the tests, they indeed made much easier to see what's going on. 
I added a patch to the PR to change the calculation for the scope of the local variable. 
It should fix the crashing issue for weird formatted tests and tests with "invisible" code, such as the handlers for the try-with-resources.  Now, the build should be successful. 
Could you please check if it works for you as well? 
Thanks, 
David

**Comment by dodie** [comment]
Awesome! I merge this PR to include the changes in the next release!
Thanks for your contribution!

**Pull request merged**